src/ParallelAlgorithms/ApparentHorizonFinder/InterpolateVolumeVars.cpp

```diff
@@ -0,0 +1,127 @@
+#include "ParallelAlgorithms/ApparentHorizonFinder/Storage.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/ErrorHandling/FloatingPointExceptions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ah {
+template <typename Fr>
+void interpolate_volume_data(
+    const gsl::not_null<ah::Storage::Iteration<Fr>*> current_iteration_storage,
+    const gsl::not_null<
+        std::unordered_map<ElementId<3>, ah::Storage::VolumeVariables<Fr>>*>
+        all_volume_variables,
+    const LinkedMessageId<double>& time, const Domain<3>& domain,
+    const domain::FunctionsOfTimeMap& functions_of_time) {
+  std::vector<ElementId<3>> element_ids;
+  element_ids.reserve(all_volume_variables->size());
+  for (const auto& [element_id, volume_vars] : *all_volume_variables) {
+    (void)volume_vars;  // Avoid clang-tidy warning
+    if (not current_iteration_storage->interpolation_is_done_for_these_elements
+                .contains(element_id)) {
+      element_ids.push_back(element_id);
+      current_iteration_storage->interpolation_is_done_for_these_elements
+          .insert(element_id);
+    }
+  }
+
+  ASSERT(current_iteration_storage->block_coord_holders.has_value(),
+         "Block logical coordinates of horizon have not been set!");
+
+  const auto& block_coord_holders =
+      current_iteration_storage->block_coord_holders.value();
+  const auto element_coord_holders =
+      element_logical_coordinates(element_ids, block_coord_holders);
+
+  for (const auto& [element_id, element_coord_holder] : element_coord_holders) {
+    const auto& offsets = element_coord_holder.offsets;
+    auto& volume_vars_storage = all_volume_variables->at(element_id);
+    auto& interpolated_vars = current_iteration_storage->interpolated_vars;
+
+    // Only fill once
+    const size_t expected_num_points = block_coord_holders.size();
+    if (interpolated_vars.number_of_grid_points() != expected_num_points) {
+      // Temporarily disable FPEs
+      const ScopedFpeState fpe_state{false};
+      interpolated_vars.initialize(
+          expected_num_points, std::numeric_limits<double>::signaling_NaN());
+    }
+
+    // Take the ah::source_vars and convert to
+    // ah::vars_to_interpolate_to_target
+    ah::compute_vars_to_interpolate_to_target(
+        make_not_null(&volume_vars_storage), time, domain, element_id,
+        functions_of_time);
+
+    const intrp::Irregular<3> interpolator(
+        volume_vars_storage.mesh, element_coord_holder.element_logical_coords);
+
+    // Vars interpolated to points within this element
+    auto local_interpolated_vars = interpolator.interpolate(
+        volume_vars_storage.vars_to_interpolate_to_target);
+
+    // Loop over each tensor
+    tmpl::for_each<ah::vars_to_interpolate_to_target<3, Fr>>([&](auto tag_v) {
+      using tag = tmpl::type_from<decltype(tag_v)>;
```

Comment:
[optional] Use an explicit template parameter on the lambda.  Also in a few other places.